models/middlewares/dbChangesLogger.js

```diff
@@ -0,0 +1,49 @@
+import logger from "@config/logger";
+
+const dbChangesLoggerMiddleware = (schema) => {
+  let username;
+  let collection;
+  let changesBefore;
+  let changesAfter;
+
+  async function beforeUpdate(next) {
+    const isUpdateOperation = this.op && this.op === "findOneAndUpdate";
+
+    username = isUpdateOperation ? this._conditions.username : this.username;
+    collection = isUpdateOperation ? this._collection.collectionName : this.constructor.collection.collectionName;
+    changesBefore = isUpdateOperation ? await this.model.findOne(this.getFilter()) : null;
+  
+    next();
+  }
+
+  async function afterUpdate(doc, next) {
+    changesAfter = await doc.constructor.findById(doc._id);
+
+    const date = new Date();
+    
+    const change = {
+      username,
+      collection,
+      collection_id: "in progress....",
+      changesBefore,
+      changesAfter,
+      date
+    };
+    
+    if (changesBefore !== changesAfter) {
+      logger.info(change);
```

Comment:
I am keen to log these changes to a new collection in Mongo

Reply:
Same here Eddie! 

But I have a certain blocker... Its weird. So..

- I have consumed the `mongoose` doc enough to know that I should to write a `Schema` then create a `Model` from the schema like so.

```js
const myNewSchema = new mongoose.Schema({ ... });

mongoose.model('MyModel', myNewSchema);
```
- I have also taken notice of style the existing models in the project are created and I have arrived at my implementation in similar fashion. _**Below is my implementation of the `DBChange` model**_ (Blocker Here)

```js
import mongoose from "mongoose";
 
const dbChangeSchema = new mongoose.Schema(
  {
    username: String,
    collection: String,
    doc: String,
    changesBefore: mongoose.Schema.Types.Mixed,
    changesAfter: mongoose.Schema.Types.Mixed,
    date: {
      type: Date,
      default: new Date(),
    }
  },
  { timestamps: true }
);

module.exports = mongoose.models.DBChange || mongoose.model("DBChange", dbChangeSchema);
```
- I hope to implement this in the `dbChangesLogger` middleware function

```js
  async function afterUpdate(doc, next) {
    ...
    const change = {
      username,
      collection,
      doc: doc._id,
      changesBefore,
      changesAfter,
      date
    };
    
    if (changesBefore !== changesAfter) {
      await DBChange.create({ ...change });   // <- CHANGE ADDED TO DB RIGHT HERE
      logger.info(change);
    } else {
      logger.info(`No changes occured in: ${collection}`)
    }

    return next();
  }
  ...
};
```

...but there goes my certain blocker.

### Blocker

- I get the below error when I try to load the website on 

**Screenshot**

![screenshot-localhost_3000-2023 08 14-20_51_55](https://github.com/EddieHubCommunity/LinkFree/assets/25631971/9067c5db-03b4-475f-b17c-3311ca7a5e4a)

**Stack Trace**
```shell
(node:2716) [MONGOOSE] Warning: `collection` is a reserved schema pathname and may break some functionality. You are allowed to use it, but use at your own risk. To disable this warning pass `suppressReservedKeysWarning` as a schema option.
(Use `node --trace-warnings ...` to show where the warning was created)
- error Error [TypeError]: Cannot read properties of undefined (reading 'Symbol(mongoose#Document#scope)')
    at Model.set [as collection] (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\mongoose\lib\helpers\document\compile.js:205:32)
    at Function.compile (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\mongoose\lib\model.js:4735:30)
    at Mongoose._model (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\mongoose\lib\index.js:620:27)
    at Mongoose.model (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\mongoose\lib\index.js:579:27)
    at eval (webpack-internal:///./models/DBChange.js:20:73)
    at ./models/DBChange.js (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:319:1)
    at __webpack_require__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:33:42)
    at eval (webpack-internal:///./models/index.js:16:67)
    at ./models/index.js (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:451:1)
    at __webpack_require__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:33:42)
    at eval (webpack-internal:///./pages/api/auth/db-adapter.js:5:71)
    at ./pages/api/auth/db-adapter.js (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:704:1)
    at __webpack_require__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:33:42)
    at eval (webpack-internal:///./pages/api/auth/[...nextauth].js:12:69)
    at ./pages/api/auth/[...nextauth].js (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:693:1)
    at __webpack_require__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:33:42)
    at eval (webpack-internal:///./pages/account/manage/links.js:9:77)
    at __webpack_require__.a (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:97:13)
    at eval (webpack-internal:///./pages/account/manage/links.js:1:21)
    at ./pages/account/manage/links.js (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:660:1)
    at __webpack_require__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:33:42)
    at eval (webpack-internal:///./node_modules/next/dist/build/webpack/loaders/next-route-loader/index.js?page=%2Faccount%2Fmanage%2Flinks&preferredRegion=&absolutePagePath=.%2Fpages%5Caccount%5Cmanage%5Clinks.js&absoluteAppPath=private-next-pages%2F_app&absoluteDocumentPath=private-next-pages%2F_document&middlewareConfigBase64=e30%3D!:23:88)
    at __webpack_require__.a (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:97:13)
    at eval (webpack-internal:///./node_modules/next/dist/build/webpack/loaders/next-route-loader/index.js?page=%2Faccount%2Fmanage%2Flinks&preferredRegion=&absolutePagePath=.%2Fpages%5Caccount%5Cmanage%5Clinks.js&absoluteAppPath=private-next-pages%2F_app&absoluteDocumentPath=private-next-pages%2F_document&middlewareConfigBase64=e30%3D!:1:21)
    at ./node_modules/next/dist/build/webpack/loaders/next-route-loader/index.js?page=%2Faccount%2Fmanage%2Flinks&preferredRegion=&absolutePagePath=.%2Fpages%5Caccount%5Cmanage%5Clinks.js&absoluteAppPath=private-next-pages%2F_app&absoluteDocumentPath=private-next-pages%2F_document&middlewareConfigBase64=e30%3D! (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:33:1)
    at __webpack_require__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\webpack-runtime.js:33:42)
    at __webpack_exec__ (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:1546:39)
    at C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:1547:28
    at Object.<anonymous> (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\.next\server\pages\account\manage\links.js:1550:3)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at requirePage (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\require.js:112:75)
    at C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\load-components.js:80:84
    at async loadComponentsImpl (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\load-components.js:80:26)
    at async DevServer.findPageComponentsImpl (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\next-server.js:772:36)
    at async DevServer.findPageComponents (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\dev\next-dev-server.js:1271:20)
    at async DevServer.renderPageComponent (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\base-server.js:1357:24)
    at async DevServer.renderToResponseImpl (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\base-server.js:1401:32)
    at async DevServer.pipeImpl (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\base-server.js:645:25)
    at async Object.fn (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\next-server.js:1153:21)
    at async Router.execute (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\router.js:315:32)
    at async DevServer.runImpl (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\base-server.js:619:29)
    at async DevServer.run (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\dev\next-dev-server.js:910:20)
    at async DevServer.handleRequestImpl (C:\Users\Bode\Documents\Projects\Open Source\LinkFree\node_modules\next\dist\server\base-server.js:546:20) {
  digest: undefined
}
```

### Other concerns

- We have some sort of `warning` in the stack trace about our usage of `collection` as property name

```shell
Warning: `collection` is a reserved schema pathname and may break some functionality. You are allowed to use it, but use at your own risk. To disable this warning pass `suppressReservedKeysWarning` as a schema option.
(Use `node --trace-warnings ...` to show where the warning was created)
```

Reply:
Can you push it to your branch so I can test it locally and try to debug

Reply:
Pushed!

Reply:
I made some tweaks and it then worked for me, here are the changes, save them to a file `8605.patch` and you can do `git apply 8605.patch`

```diff
diff --git a/models/Changelog.js b/models/Changelog.js
new file mode 100644
index 00000000..e3adb591
--- /dev/null
+++ b/models/Changelog.js
@@ -0,0 +1,15 @@
+import mongoose from "mongoose";
+
+const ChangelogSchema = new mongoose.Schema(
+  {
+    username: String,
+    collectionId: String,
+    collectionName: String,
+    changesBefore: mongoose.Schema.Types.Mixed,
+    changesAfter: mongoose.Schema.Types.Mixed,
+  },
+  { timestamps: true }
+);
+
+module.exports =
+  mongoose.models.Changelog || mongoose.model("Changelog", ChangelogSchema);
diff --git a/models/DBChange.js b/models/DBChange.js
deleted file mode 100644
index 3dd4c487..00000000
--- a/models/DBChange.js
+++ /dev/null
@@ -1,18 +0,0 @@
-import mongoose from "mongoose";
- 
-const dbChangeSchema = new mongoose.Schema(
-  {
-    username: String,
-    collection: String,
-    doc: String,
-    changesBefore: mongoose.Schema.Types.Mixed,
-    changesAfter: mongoose.Schema.Types.Mixed,
-    date: {
-      type: Date,
-      default: new Date(),
-    }
-  },
-  { timestamps: true }
-);
-
-module.exports = mongoose.models.DBChange || mongoose.model("DBChange", dbChangeSchema);
\ No newline at end of file
diff --git a/models/index.js b/models/index.js
index 5b9f86fa..79b4a2ff 100644
--- a/models/index.js
+++ b/models/index.js
@@ -1,5 +1,5 @@
 import Account from "./Account";
-import DBChange from "./DBChange";
+import Changelog from "./Changelog";
 import Link from "./Link";
 import LinkStats from "./LinkStats";
 import Profile from "./Profile";
@@ -11,7 +11,7 @@ import VerificationToken from "./VerificationToken";
 
 export {
   Account,
-  DBChange,
+  Changelog,
   Link,
   LinkStats,
   Profile,
diff --git a/models/middlewares/dbChangesLogger.js b/models/middlewares/dbChangesLogger.js
index 0ceba02c..0cf1b86d 100644
--- a/models/middlewares/dbChangesLogger.js
+++ b/models/middlewares/dbChangesLogger.js
@@ -1,5 +1,5 @@
 import logger from "@config/logger";
-import { DBChange } from "@models/index";
+import { Changelog } from "@models/index";
 
 const dbChangesLoggerMiddleware = (schema) => {
   let username;
@@ -11,9 +11,13 @@ const dbChangesLoggerMiddleware = (schema) => {
     const isUpdateOperation = this.op && this.op === "findOneAndUpdate";
 
     username = isUpdateOperation ? this._conditions.username : this.username;
-    collection = isUpdateOperation ? this._collection.collectionName : this.constructor.collection.collectionName;
-    changesBefore = isUpdateOperation ? await this.model.findOne(this.getFilter()) : null;
-  
+    collection = isUpdateOperation
+      ? this._collection.collectionName
+      : this.constructor.collection.collectionName;
+    changesBefore = isUpdateOperation
+      ? await this.model.findOne(this.getFilter())
+      : null;
+
     next();
   }
 
@@ -21,26 +25,25 @@ const dbChangesLoggerMiddleware = (schema) => {
     changesAfter = await doc.constructor.findById(doc._id);
 
     const date = new Date();
-    
+
     const change = {
       username,
-      collection,
-      doc: doc._id,
+      collectionName: collection,
+      collectionId: doc._id,
       changesBefore,
       changesAfter,
-      date
     };
-    
+
     if (changesBefore !== changesAfter) {
-      await DBChange.create({ ...change });
+      await Changelog.create({ ...change });
       logger.info(change);
     } else {
-      logger.info(`No changes occured in: ${collection}`)
+      logger.info(`No changes occured in: ${collection}`);
     }
 
     return next();
   }
-  
+
   schema.pre("save", beforeUpdate);
   schema.pre("findOneAndUpdate", beforeUpdate);
```

Reply:
Works beautifully @eddiejaoude 

Haha, I will just take it that the `warning` about the `collection` reserved term usage indeed broke some functionality as mentioned in there,

```shell
Warning: `collection` is a reserved schema pathname and may break some functionality. You are allowed to use it, but use at your own risk. To disable this warning pass `suppressReservedKeysWarning` as a schema option.
(Use `node --trace-warnings ...` to show where the warning was created)
```

...considering there wasn't lots of changes beyond the rename context. 

### Notable Learning

- Now I see what the `{ timestamps: true }` does in the schema, I see it adds the `createdAt` and `upatedAt` timestamp properties.

### Other Thoughts

- I see that you have opted for the term `collectionId` to hold the `doc._id`. I had some thoughts concerning the property naming... _see comment_ https://github.com/EddieHubCommunity/LinkFree/issues/7660#issuecomment-1675863310

😉 PS: I wasn't able to get the changes apply via `git apply <.patch file>` command due to some `trailing whitespace.` and `corrupt patch at line***` error. I tried to fix the patch using the `--whitespace=fix` flag, but no error persists, I had to manually commit the changes.